source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -14,41 +17,95 @@ DnsFilterEnvoyConfig::DnsFilterEnvoyConfig(
+    const auto& dns_table = server_config.inline_dns_table();
+    const size_t entries = dns_table.virtual_domains().size();
 
-    // TODO (abaptiste): Check that the domain configured here appears
-    // in the known domains list
     virtual_domains_.reserve(entries);
-    for (const auto& virtual_domain : cfg.virtual_domains()) {
-      DnsAddressList addresses{};
+    for (const auto& virtual_domain : dns_table.virtual_domains()) {
+      AddressConstPtrVec addrs{};
 
       if (virtual_domain.endpoint().has_address_list()) {
         const auto& address_list = virtual_domain.endpoint().address_list().address();
-        addresses.reserve(address_list.size());
+        addrs.reserve(address_list.size());
+        // This will throw an exception if the configured_address string is malformed
         for (const auto& configured_address : address_list) {
-          addresses.push_back(configured_address);
+          const auto ipaddr = Network::Utility::parseInternetAddress(
+              configured_address, 0 /* port */, true /* v6only */);
+          addrs.push_back(ipaddr);
         }
       }
+      virtual_domains_.emplace(virtual_domain.name(), std::move(addrs));
+
+      uint64_t ttl = virtual_domain.has_answer_ttl()
+                         ? DurationUtil::durationToSeconds(virtual_domain.answer_ttl())
+                         : DefaultResolverTTLs;
+      domain_ttl_.emplace(virtual_domain.name(), ttl);
+    }
 
-      virtual_domains_.emplace(virtual_domain.name(), std::move(addresses));
+    // Add known domains
+    known_suffixes_.reserve(dns_table.known_suffixes().size());
+    for (const auto& suffix : dns_table.known_suffixes()) {
+      // TODO(abaptiste): We support only suffixes here. Expand this to support other StringMatcher
+      // types
+      envoy::type::matcher::v3::StringMatcher matcher;
+      matcher.set_suffix(suffix.suffix());
+      auto matcher_ptr = std::make_unique<Matchers::StringMatcherImpl>(matcher);
+      known_suffixes_.push_back(std::move(matcher_ptr));
     }
   }
+
+  const auto& client_config = config.client_config();
+  forward_queries_ = client_config.forward_query();
+  if (forward_queries_) {
+    const auto& upstream_resolvers = client_config.upstream_resolvers();
+    resolvers_.reserve(upstream_resolvers.size());
+    for (const auto& resolver : upstream_resolvers) {
+      auto ipaddr =
+          Network::Utility::parseInternetAddress(resolver, 0 /* port */, true /* v6only */);
+      resolvers_.push_back(std::move(ipaddr));
+    }
+  }
+
+  resolver_timeout_ms_ = std::chrono::milliseconds(
+      PROTOBUF_GET_MS_OR_DEFAULT(client_config, resolver_timeout, DefaultResolverTimeoutMs));
 }
 
 void DnsFilter::onData(Network::UdpRecvData& client_request) {
-  // Handle incoming request and respond with an answer
-  UNREFERENCED_PARAMETER(client_request);
+
+  // Save the connection endpoints so that we can respond
+  local_ = client_request.addresses_.local_;
```

Comment:
Yes.  Primarily because of the asynchronous external DNS resolution, the local and peer address need to be class members to construct the reply.

Reply:
Not necessarily. You could capture it in a lambda which has the request context and retrieve it from there after the async work is done. It would keep your class cleaner

Reply:
Also - when you actually get to implementing this async external query logic, you will see what I mean. This is just based on past experience of having done this.